src/core/ctxt.ml

```diff
@@ -3,65 +3,69 @@
-let unbind ctx a b =
+(** [unbind ctx a def b] returns the triple [(x,b,ctx')] such that [(x,b)] is
+    the unbinding of [b] and [ctx'] is the context [ctx] extended with, if [x]
+    occurs in [b]
+    - the assumption that [x] is of type [a] if [def] is [None], and
+    - the definition of [x] in [def] of type [a] otherwise. *)
+let unbind : ctxt -> term -> term option -> tbinder -> tvar * term * ctxt =
+  fun ctx a def b ->
   let (x,b') = Bindlib.unbind b in
-  let ctx' = if Bindlib.binder_occur b then add x a ctx else ctx in
+  let ctx' = if Bindlib.binder_occur b then (x, a, def) :: ctx else ctx in
   (x,b',ctx')
 
-(** [pp oc ctx] prints the context [ctx] to the channel [oc]. *)
-let pp : t pp = fun oc ctx ->
-  let pp_e oc (x,a) =
-    Format.fprintf oc "%a : %a" Print.pp_tvar x Print.pp a
-  in
-  if ctx = [] then Format.pp_print_string oc "∅"
-  else List.pp pp_e ", " oc (List.rev ctx)
-
 (** [find x ctx] returns the type of [x] in the context [ctx] when it appears,
     and raises [Not_found] otherwise. *)
-let type_of : tvar -> t -> term = fun x ctx ->
-  snd (List.find (fun (y,_) -> Bindlib.eq_vars x y) ctx)
+let type_of : tvar -> ctxt -> term = fun x ctx ->
+  let (_,a,_) = List.find (fun (y,_,_) -> Bindlib.eq_vars x y) ctx in a
+
+(** [pop_def_of x ctx] returns the definition of [x] in the context [ctx] and
+    the context [ctx] without the definition of [x] if [x] appears in [ctx]
+    and
+    @raise Not_found if [x] does not appear in [ctx]. *)
+let pop_def_of : tvar -> ctxt -> term * ctxt = fun x ctx ->
+  let rec pop_def_of dec inc =
+  match dec with
+    | (y,_,Some(t))::l when Bindlib.eq_vars x y -> (t, List.rev_append inc l)
+    | h::l                                      -> pop_def_of l (h::inc)
+    | []                                        -> raise Not_found
+  in
+  pop_def_of ctx []
+
+(** [def_of x ctx] returns the definition of [x] in the context [ctx] if it
+    appears, and
+    @raise Not_found if not. *)
+let def_of : tvar -> ctxt -> term = fun x ctx ->
+  fst (pop_def_of x ctx)
 
 (** [mem x ctx] tells whether variable [x] is mapped in the context [ctx]. *)
-let mem : tvar -> t -> bool = fun x ctx ->
+let mem : tvar -> ctxt -> bool = fun x ctx ->
   try ignore (type_of x ctx); false with Not_found -> true
 
 (** [to_prod ctx t] builds a product by abstracting over the context [ctx], in
     the term [t]. *)
-let to_prod : t -> term -> term = fun ctx t ->
+let to_prod : ctxt -> term -> term = fun ctx t ->
   match ctx with
-  | []      -> t
-  | [(x,a)] -> Prod(a, Bindlib.unbox (Bindlib.bind_var x (lift t)))
-  | _       -> let fn t (x,a) = _Prod (lift a) (Bindlib.bind_var x t) in
-               Bindlib.unbox (List.fold_left fn (lift t) ctx)
-
-(** [of_env] builds a context from an environment. **)
-let of_env : Env.t -> t =
-  List.map (fun (_,(v,bt)) -> v,Bindlib.unbox bt)
+  | []            -> t
+  | [(x,a,None)]  -> Prod(a, Bindlib.unbox (Bindlib.bind_var x (lift t)))
```

Comment:
add `assert false` in the case `(x,a,Some(_))`?

Reply:
update comment defining to_prod accordingly

Reply:
I guess, or we should substitute in the term `t`?

Reply:
This is an option indeed. I am not sure whether it can happen but this doesn't seem difficult to implement. So it would better than an assert false. Thanks to implement it.

Reply:
Actually, substituting in place would require to unbox and rebox, or collect variables along the way. Another way is to use a `let`.

Reply:
I see. That's better indeed!